LSP in VSCode does not work

### Informations about my hardware and software:

- openHAB server: xen vm 64bit 1GByte RAM, 1 Processor
  - os: debian stretch 64bit
  - java: Oracle 
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)
  - openHAB: openHAB Distribution Version Information
build-no        : Build #1088
online-repo     : https://openhab.jfrog.io/openhab/online-repo-snapshot/2.2
Repository        Version
openhab-distro  : 2.2.0-SNAPSHOT
smarthome       : 0.9.0-SNAPSHOT
openhab-core    : 2.2.0-SNAPSHOT
openhab1-addons : 1.11.0-SNAPSHOT
openhab2-addons : 2.2.0-SNAPSHOT
karaf           : 4.1.3
  - samba: Samba version 4.5.12-Debian
Protcoll is SMB2_10
- VSCode Client: Windows 7 professional 64bit, 8GByte RAM
  - VSCode: 
Version 1.18.1
Commit 929bacba01ef658b873545e26034d1a8067445e9
Datum 2017-11-16T18:32:36.023Z
Shell 1.7.9
Renderer 58.0.3029.110
Node 7.9.0
Architektur x64
  - openHAB plugin: 0.3.0

When opening the conf folder by using the URI (`\\fqdn\openhab-conf\`) vscode gives an ERROR "Internal error, please look at the server's logs."

Server log is: 
```
11:18:40.037 [ERROR] [.eclipse.lsp4j.jsonrpc.RemoteEndpoint] - Internal error: java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:53) ~[102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:105) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:203) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:139) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:149) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:77) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:84) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:51) ~[?:?]
        ... 11 more
Caused by: java.lang.IllegalArgumentException: URI has an authority component
        at sun.nio.fs.UnixUriUtils.fromUri(UnixUriUtils.java:53) ~[?:?]
        at sun.nio.fs.UnixFileSystemProvider.getPath(UnixFileSystemProvider.java:98) ~[?:?]
        at java.nio.file.Paths.get(Paths.java:138) ~[?:?]
        at org.eclipse.smarthome.model.lsp.internal.MappingUriExtensions.guessClientPath(MappingUriExtensions.java:94) ~[?:?]
        at org.eclipse.smarthome.model.lsp.internal.MappingUriExtensions.toUri(MappingUriExtensions.java:66) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.getBaseDir(LanguageServerImpl.java:225) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.initialize(LanguageServerImpl.java:176) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:51) ~[?:?]
        ... 11 more
```
Most interesting part is this:
```
Caused by: java.lang.IllegalArgumentException: URI has an authority component
```
Certainly the samba share is set to use user and password authentication.

When using a mapped windows share (`Q:\`, is set to `\\fqdn\openhab-conf\`) the VSCode ERROR stays the same, but the log is slightly another:
```
11:45:03.192 [INFO ] [del.lsp.internal.MappingUriExtensions] - Path mapping could not be done for 'file:///q%3A/', leaving it untouched
11:45:03.222 [ERROR] [.eclipse.lsp4j.jsonrpc.RemoteEndpoint] - Internal error: java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:53) ~[102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:105) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:203) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:139) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:149) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:77) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:84) [102:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:51) ~[?:?]
        ... 11 more
Caused by: java.lang.NullPointerException
        at java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011) ~[?:?]
        at java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1006) ~[?:?]
        at org.eclipse.xtext.resource.impl.ChunkedResourceDescriptions.setContainer(ChunkedResourceDescriptions.java:121) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.lambda$createFreshResourceSet$5(ProjectManager.java:152) ~[?:?]
        at org.eclipse.xtext.xbase.lib.ObjectExtensions.operator_doubleArrow(ObjectExtensions.java:139) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.createFreshResourceSet(ProjectManager.java:155) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.lambda$newBuildRequest$4(ProjectManager.java:130) ~[?:?]
        at org.eclipse.xtext.xbase.lib.ObjectExtensions.operator_doubleArrow(ObjectExtensions.java:139) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.newBuildRequest(ProjectManager.java:141) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.doBuild(ProjectManager.java:111) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.doInitialBuild(ProjectManager.java:107) ~[?:?]
        at org.eclipse.xtext.ide.server.BuildManager.doInitialBuild(BuildManager.java:138) ~[?:?]
        at org.eclipse.xtext.ide.server.WorkspaceManager.refreshWorkspaceConfig(WorkspaceManager.java:142) ~[?:?]
        at org.eclipse.xtext.ide.server.WorkspaceManager.initialize(WorkspaceManager.java:113) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.lambda$initialize$3(LanguageServerImpl.java:213) ~[?:?]
        at org.eclipse.xtext.ide.server.concurrent.RequestManager.runWrite(RequestManager.java:71) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.initialize(LanguageServerImpl.java:216) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:51) ~[?:?]
        ... 11 more
```
Most interesting part is now:
```
11:45:03.192 [INFO ] [del.lsp.internal.MappingUriExtensions] - Path mapping could not be done for 'file:///q%3A/', leaving it untouched
```
and therefor there is a null pointer exception, because there is no path set at all.

--- a/bundles/model/org.eclipse.smarthome.model.lsp.test/.classpath
+++ b/bundles/model/org.eclipse.smarthome.model.lsp.test/.classpath
@@ -1,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src/test/java"/>
+	<classpathentry kind="output" path="target/classes"/>
+</classpath>

--- a/bundles/model/org.eclipse.smarthome.model.lsp.test/.project
+++ b/bundles/model/org.eclipse.smarthome.model.lsp.test/.project
@@ -1,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>org.eclipse.smarthome.model.lsp.test</name>
+	<comment></comment>
+	<projects>
+	</projects>
+	<buildSpec>
+		<buildCommand>
+			<name>org.eclipse.jdt.core.javabuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.pde.ManifestBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.pde.SchemaBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.pde.PluginNature</nature>
+		<nature>org.eclipse.jdt.core.javanature</nature>
+	</natures>
+</projectDescription>

--- a/bundles/model/org.eclipse.smarthome.model.lsp.test/.settings/org.eclipse.jdt.groovy.core.prefs
+++ b/bundles/model/org.eclipse.smarthome.model.lsp.test/.settings/org.eclipse.jdt.groovy.core.prefs
@@ -1,0 +1,2 @@
+eclipse.preferences.version=1
+groovy.compiler.level=-1

--- a/bundles/model/org.eclipse.smarthome.model.lsp.test/META-INF/MANIFEST.MF
+++ b/bundles/model/org.eclipse.smarthome.model.lsp.test/META-INF/MANIFEST.MF
@@ -1,0 +1,20 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Eclipse SmartHome Language Server Tests
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-SymbolicName: org.eclipse.smarthome.model.lsp.test
+Bundle-Version: 0.9.0.qualifier
+Fragment-Host: org.eclipse.smarthome.model.lsp
+Import-Package: 
+ org.eclipse.jdt.annotation;resolution:=optional,
+ org.eclipse.smarthome.test,
+ org.eclipse.smarthome.test.java,
+ org.eclipse.smarthome.test.storage,
+ org.hamcrest;core=split,
+ org.junit;version="4.0.0",
+ org.mockito,
+ org.slf4j
+Require-Bundle: 
+ org.hamcrest,
+ org.junit,
+ org.mockito

--- a/bundles/model/org.eclipse.smarthome.model.lsp.test/about.html
+++ b/bundles/model/org.eclipse.smarthome.model.lsp.test/about.html
@@ -1,0 +1,28 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
+    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml">
+<head>
+<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"/>
+<title>About</title>
+</head>
+<body lang="EN-US">
+<h2>About This Content</h2>
+ 
+<p>June 5, 2006</p>	
+<h3>License</h3>
+
+<p>The Eclipse Foundation makes available all content in this plug-in (&quot;Content&quot;).  Unless otherwise 
+indicated below, the Content is provided to you under the terms and conditions of the
+Eclipse Public License Version 1.0 (&quot;EPL&quot;).  A copy of the EPL is available 
+at <a href="http://www.eclipse.org/legal/epl-v10.html">http://www.eclipse.org/legal/epl-v10.html</a>.
+For purposes of the EPL, &quot;Program&quot; will mean the Content.</p>
+
+<p>If you did not receive this Content directly from the Eclipse Foundation, the Content is 
+being redistributed by another party (&quot;Redistributor&quot;) and different terms and conditions may
+apply to your use of any object code in the Content.  Check the Redistributor's license that was 
+provided with the Content.  If no such license exists, contact the Redistributor.  Unless otherwise
+indicated below, the terms and conditions of the EPL still apply to any source code in the Content
+and such source code may be obtained at <a href="http://www.eclipse.org/">http://www.eclipse.org</a>.</p>
+
+</body>
+</html>

--- a/bundles/model/org.eclipse.smarthome.model.lsp.test/build.properties
+++ b/bundles/model/org.eclipse.smarthome.model.lsp.test/build.properties
@@ -1,0 +1,4 @@
+source.. = src/test/java
+bin.includes = META-INF/,\
+               .,\
+               about.html

--- a/bundles/model/org.eclipse.smarthome.model.lsp.test/pom.xml
+++ b/bundles/model/org.eclipse.smarthome.model.lsp.test/pom.xml
@@ -1,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+  <parent>
+    <groupId>org.eclipse.smarthome.bundles</groupId>
+    <artifactId>model</artifactId>
+    <version>0.9.0-SNAPSHOT</version>
+  </parent>
+
+
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>org.eclipse.smarthome.model</groupId>
+  <artifactId>org.eclipse.smarthome.model.lsp.test</artifactId>
+
+  <name>Eclipse SmartHome Language Server Tests</name>
+
+  <packaging>eclipse-test-plugin</packaging>
+
+</project>

--- a/bundles/model/org.eclipse.smarthome.model.lsp.test/src/test/java/org/eclipse/smarthome/model/lsp/internal/MappingUriExtensionsTest.java
+++ b/bundles/model/org.eclipse.smarthome.model.lsp.test/src/test/java/org/eclipse/smarthome/model/lsp/internal/MappingUriExtensionsTest.java
@@ -1,0 +1,137 @@
+/**
+ * Copyright (c) 2014,2017 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.model.lsp.internal;
+
+import static org.junit.Assert.assertEquals;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collection;
+
+import org.eclipse.emf.common.util.URI;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ *
+ * @author Simon Kaufmann - initial contribution and API.
+ *
+ */
+@RunWith(Parameterized.class)
+public class MappingUriExtensionsTest {
+
+    @Rule
+    public final TemporaryFolder folder = new TemporaryFolder();
+    private File confFolder;
+    private File itemsFolder;
+    private File itemsFile;
+    private final String conf;
+    private final String request;
+    private final String expectedClientPath;
+    private final String expectedUriPath;
+
+    @Parameters
+    public static Collection<Object[]> data() {
+        return Arrays.asList(new Object[][] { //
+                { "conf", //
+                        "file:///q:/conf", //
+                        "file:///q:/conf", //
+                        "/" }, //
+                { "conf", //
+                        "file:///q:", //
+                        "file:///q:", //
+                        "/" }, //
+                { "conf", //
+                        "file://fqdn/openhab-conf", //
+                        "file://fqdn/openhab-conf", //
+                        "/" }, //
+                { "conf", //
+                        "file://fqdn/openhab-conf/conf", //
+                        "file://fqdn/openhab-conf/conf", //
+                        "/" }, //
+                { null, //
+                        "file:///asdf/conf/items/test.items", //
+                        "file:///asdf/conf", //
+                        "/items/test.items" }, //
+                { null, //
+                        "file:///asdf/items/test.items", //
+                        "file:///asdf", //
+                        "/items/test.items" }, //
+                { null, //
+                        "file://fqdn/openhab-conf/conf/items/test.items", //
+                        "file://fqdn/openhab-conf/conf", //
+                        "/items/test.items" }, //
+                { null, //
+                        "file://fqdn/openhab-conf/items/test.items", //
+                        "file://fqdn/openhab-conf", //
+                        "/items/test.items" },//
+        });
+    }
+
+    public MappingUriExtensionsTest(String conf, String request, String expectedClientPath, String expectedUriPath) {
+        this.conf = conf;
+        this.request = request;
+        this.expectedClientPath = expectedClientPath;
+        this.expectedUriPath = expectedUriPath;
+    }
+
+    @Before
+    public void setup() throws Exception {
+        confFolder = folder.newFolder("conf");
+        itemsFolder = new File(confFolder, "items");
+        itemsFolder.mkdirs();
+        itemsFile = new File(itemsFolder, "test.items");
+        itemsFile.deleteOnExit();
+        itemsFile.createNewFile();
+    }
+
+    @Test
+    public void testGuessClientPath() {
+        MappingUriExtensions mapper = createMapper();
+        String clientPath = mapper.guessClientPath(request);
+        assertEquals(expectedClientPath, clientPath);
+    }
+
+    @Test
+    public void testToUri() {
+        MappingUriExtensions mapper = createMapper();
+        URI clientPath = mapper.toUri(request);
+        assertEquals("file://" + confFolder.getAbsolutePath() + expectedUriPath, clientPath.toString());
+    }
+
+    @Test
+    public void testToPath_emfURI() {
+        MappingUriExtensions mapper = createMapper();
+        mapper.toUri(request);
+
+        URI uri = URI.createURI("file://" + confFolder.getAbsolutePath() + expectedUriPath);
+        String res = mapper.toPath(uri);
+        assertEquals(request, res);
+    }
+
+    private MappingUriExtensions createMapper() {
+        return new MappingUriExtensions(conf) {
+            @Override
+            protected String calcServerLocation(String configFolder) {
+                // ensure test execution is independent from the current working directory
+                return "file://" + confFolder.getAbsolutePath();
+            }
+        };
+    }
+
+}

--- a/bundles/model/org.eclipse.smarthome.model.lsp/src/main/java/org/eclipse/smarthome/model/lsp/internal/MappingUriExtensions.java
+++ b/bundles/model/org.eclipse.smarthome.model.lsp/src/main/java/org/eclipse/smarthome/model/lsp/internal/MappingUriExtensions.java
@@ -13,6 +13,10 @@
 package org.eclipse.smarthome.model.lsp.internal;
 
 import java.io.File;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.FileSystemNotFoundException;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -33,17 +37,25 @@ import org.slf4j.LoggerFactory;
  */
 public class MappingUriExtensions extends UriExtensions {
 
+    private static final Charset PATH_ENCODING = StandardCharsets.UTF_8;
+
     private final Logger logger = LoggerFactory.getLogger(MappingUriExtensions.class);
 
     private final String serverLocation;
     private String clientLocation = null;
+    private final String rawConfigFolder;
 
     public MappingUriExtensions(String configFolder) {
+        this.rawConfigFolder = configFolder;
+        serverLocation = calcServerLocation(configFolder);
+        logger.debug("The language server is using '{}' as its workspace", serverLocation);
+    }
+
+    protected String calcServerLocation(String configFolder) {
         Path configPath = Paths.get(configFolder);
         Path absoluteConfigPath = configPath.toAbsolutePath();
         java.net.URI configPathURI = absoluteConfigPath.toUri();
-        serverLocation = configPathURI.toString();
-        logger.debug("The language server is using '{}' as its workspace", serverLocation);
+        return removeTrailingSlash(configPathURI.toString());
     }
 
     @Override
@@ -59,30 +71,44 @@ public class MappingUriExtensions extends UriExtensions {
         } catch (FileSystemNotFoundException e) {
             // fall-back to the argument
         }
-        return ret.toString().replace(serverLocation, clientLocation);
+        String clientPath = removeTrailingSlash(ret.toASCIIString());
+        if (clientLocation != null) {
+            clientPath = clientPath.replace(serverLocation, clientLocation);
+        }
+        return clientPath;
     }
 
     @Override
     public URI toUri(String pathWithScheme) {
-        if (clientLocation != null && pathWithScheme.startsWith(clientLocation)) {
-            return map(pathWithScheme, clientLocation);
-        }
+        try {
+            String decodedPathWithScheme = URLDecoder.decode(pathWithScheme, PATH_ENCODING.toString());
 
-        clientLocation = guessClientPath(pathWithScheme);
-        if (clientLocation != null) {
-            logger.debug("Identified client workspace as '{}'", clientLocation);
-            return map(pathWithScheme, clientLocation);
-        }
+            if (clientLocation != null && decodedPathWithScheme.startsWith(clientLocation)) {
+                return map(decodedPathWithScheme);
+            }
 
-        logger.info("Path mapping could not be done for '{}', leaving it untouched", pathWithScheme);
+            clientLocation = guessClientPath(decodedPathWithScheme);
+            if (clientLocation != null) {
+                logger.debug("Identified client workspace as '{}'", clientLocation);
+                return map(decodedPathWithScheme);
+            }
+
+            clientLocation = pathWithScheme;
+
+        } catch (UnsupportedEncodingException e) {
+            logger.error("Charset {} is not supported. You're seriously in trouble.", PATH_ENCODING);
+        }
+        logger.debug("Path mapping could not be done for '{}', leaving it untouched", pathWithScheme);
         java.net.URI javaNetUri = java.net.URI.create(pathWithScheme);
         return URI.createURI(super.toPath(javaNetUri));
     }
 
-    private String toURIString(String filename) {
-        Path path = Paths.get(filename);
-        java.net.URI pathURI = path.toUri();
-        return pathURI.toString();
+    private String removeTrailingSlash(String path) {
+        if (path.endsWith("/")) {
+            return path.substring(0, path.length() - 1);
+        } else {
+            return path;
+        }
     }
 
     /**
@@ -96,23 +122,50 @@ public class MappingUriExtensions extends UriExtensions {
      * @return the substring which needs to be replaced with the runtime's config folder path
      */
     protected String guessClientPath(String pathWithScheme) {
-        File file = Paths.get(java.net.URI.create(pathWithScheme)).toFile().getParentFile();
-        File realFile;
-        while (file.getParentFile() != null) {
-            String toReplace = toURIString(file.toString()) + File.separator;
-            String path = pathWithScheme.replace(toReplace, serverLocation);
-            realFile = new File(java.net.URI.create(path));
-            if (realFile.exists()) {
-                return toReplace;
-            }
-            file = file.getParentFile();
+        if (isPointingToConfigFolder(pathWithScheme)) {
+            return removeTrailingSlash(pathWithScheme);
+        } else if (isFolder(pathWithScheme)) {
+            return removeTrailingSlash(pathWithScheme);
         }
+
+        String currentPath = pathWithScheme;
+        int nextIndex = getLastPathSegmentIndex(currentPath);
+        while (nextIndex > -1) {
+            currentPath = currentPath.substring(0, nextIndex);
+            java.net.URI uri = toURI(pathWithScheme, currentPath);
+            File realFile = new File(uri);
+            if (realFile.exists()) {
+                return currentPath;
+            }
+
+            nextIndex = getLastPathSegmentIndex(currentPath);
+        }
+
         return null;
     }
 
-    private URI map(String pathWithScheme, String clientLocation) {
-        java.net.URI javaNetUri = java.net.URI.create(pathWithScheme.replace(clientLocation, serverLocation));
-        return URI.createURI(super.toPath(javaNetUri));
+    private boolean isFolder(String currentPath) {
+        return !currentPath.substring(getLastPathSegmentIndex(currentPath)).contains(".");
+    }
+
+    private boolean isPointingToConfigFolder(String currentPath) {
+        return currentPath.endsWith("/" + rawConfigFolder);
+    }
+
+    private int getLastPathSegmentIndex(String currentPath) {
+        return removeTrailingSlash(currentPath).lastIndexOf("/");
+    }
+
+    private URI map(String pathWithScheme) {
+        java.net.URI javaNetUri = toURI(pathWithScheme, clientLocation);
+        logger.trace("Going to map path {}", javaNetUri);
+        URI ret = URI.createURI(super.toPath(javaNetUri));
+        logger.trace("Mapped path {} to {}", pathWithScheme, ret);
+        return ret;
+    }
+
+    private java.net.URI toURI(String pathWithScheme, String currentPath) {
+        return java.net.URI.create(pathWithScheme.replace(currentPath, serverLocation));
     }
 
 }

--- a/bundles/model/pom.xml
+++ b/bundles/model/pom.xml
@@ -25,6 +25,7 @@
     <module>org.eclipse.smarthome.model.item.ui</module>
     <module>org.eclipse.smarthome.model.item.tests</module>
     <module>org.eclipse.smarthome.model.lsp</module>
+    <module>org.eclipse.smarthome.model.lsp.test</module>
     <module>org.eclipse.smarthome.model.persistence</module>
     <module>org.eclipse.smarthome.model.persistence.ide</module>
     <module>org.eclipse.smarthome.model.persistence.runtime</module>

